dvc/fs/fsspec_wrapper.py

```diff
@@ -91,7 +91,17 @@ def copy(self, from_info, to_info):
         self.fs.copy(self._with_bucket(from_info), self._with_bucket(to_info))
 
     def exists(self, path_info, use_dvcignore=False):
-        return self.fs.exists(self._with_bucket(path_info))
+        # Some implementations don't use ls under the hood
+        # which is not an efficient behavior on DVC considering that
+        # we usually to a lot of stat calls. For increasing the cache
+        # construction, we will first try to see whether that directory
+        # exists by doing an ls() on it's parent, and if that fails we
+        # would fall back to the original implementation.
+        raw_path = self._with_bucket(path_info)
+        try:
+            return raw_path in self.fs.ls(self._with_bucket(path_info.parent))
```

Comment:
hmm ... this feels like a very expensive op compared to a regular HEAD call (I would expect that's what `exists` do). Could you please clarify why is it needed? It was hard for me to understand it from the comment.

Reply:
Actually I'd say it is not that expensive (considering that the pagination works asyncronously). If the parent directory consists from 1000< items, then there is not much of an overhead. But if it has 8K items then the cost is about 3.5X worse for a single time. 

Considering that we do exists() calls for the same objects sometimes 2 sometimes even 3 (accomponied by .isfile()/.isdir()) I think it worths it. I am not really sure whether I should only apply this optimization to s3 implementation or in general, will actually intend to run benchmarks for the other remotes we have (adlfs/gcsfs) [currently still running more s3 benchmarks]). But I think it is a nice one (at least for s3)

Reply:
Hmm ... I'm trying to wrap my mind around it :)

1. Okay, 8K works, but what if we have 1M?
2. Transferring even 1000 results is probably 1000x slower/ takes 1000x traffic? Means that in a slower network connection situation we might hit a bigger performance problem?
3. `Considering that we do exists() calls for the same objects sometimes 2 sometimes even 3` - can we improve/fix this? How the current approach helps us there. May be I'm missing something.


To be honest, it feels that I'm missing some important information here. Could explain me the comment about the `ls`? Let's put it this way - I'm not sure I understand why it's an optimization at all.

Reply:
> To be honest, it feels that I'm missing some important information here.

fsspec basically caches all the directory listings and later use those stuff on different calls (like info()/ls()/exists() etc). What we are basically doing here is sort of pre-populating the cache with the assumption of we will use later. Not sure how much I am missing, but it always shows a positive performance on my tests. 

> Okay, 8K works, but what if we have 1M?
> Transferring even 1000 results is probably 1000x slower/ takes 1000x traffic? Means that in a slower network connection situation we might hit a bigger performance problem?

Right, these cases seem problematic. Will think more about this one (maybe separate it out)

Reply:
> fsspec basically caches all the directory listings and later use those stuff on different calls 

interesting! how/when do they invalidate this cache? 

> What we are basically doing here is sort of pre-populating the cache with the assumption of we will use later.

gotcha. Thanks. It makes sense. I wonder if we can do optimize this with a simple memoize decorator (I think `funcy` has it). But then we'll have the same question - when/how do we invalidate it? It can be a quite strong assumption and an important behavior detail since I can easily imagine where in certain cases we actually want to go and check an object again.

Reply:
Basically every operation (e.g copy, move, rm, upload) invalidates the cache for that directory, e.g;
https://github.com/iterative/dvc/blob/a3a55339abf133e8073a5c4220e7d61749b5f16f/dvc/fs/fsspec_wrapper.py#L141

By the way, I decided to drop this optimization for this PR. Instead I might go with #5689 which would definietly improve the amount of calls we do to exists() + info() (both in the master and then in this PR).

Reply:
Yep, thanks, @isidentical . It's still really good to know about this built-in cache mechanism anyways. Since DVC is used in a distributed environment it might happen that it won't detect certain files when it "expects" them? I hope that it will be just a matter of some non-optimal behavior in some rare edge cases (e.g. dir uploaded twice, or something).

I wonder btw, why don't they just cache `exists` calls as well?

Reply:
> Since DVC is used in a distributed environment it might happen that it won't detect certain files when it "expects" them?

Unlike the internal database we use to hold stuff (diskcache), this is a per-process cache that is hold in memory so I'd except the amount of race conditions to be minimum (though it can still happen).

Reply:
> I wonder btw, why don't they just cache exists calls as well?

The structure they use for cache (dircache) is in the form of a directory where you'd keep `{'dir': [{'name': 'foo'}, {'name': 'bar'}]}` etc. So from what I can presume, individual items doesn't fit (AFAIK they can't simply create a new entry on the dircache without knowing all the contents, since the code assumes the dircache is complete and raises FileNotFoundError etc. when it can't find something over there).